_source/_posts/2022-08-19-build-crud-spring-and-vue.md

```diff
@@ -0,0 +1,1057 @@
+        <q-btn flat round dense icon="account_circle" v-else @click="login"/>
+      </q-toolbar>
+    </q-header>
+
+    <q-page-container>
+      <router-view></router-view>
+    </q-page-container>
+
+  </q-layout>
+</template>
+
+<script>
+export default {
+  name: 'LayoutDefault',
+  data: function () {
+    return {
+      claims: null
+    }
+  },
+  watch: {
+    'authState.isAuthenticated'() {
+      this.$log.debug(("watch triggered!"))
+      this.updateClaims()
+    }
+  },
+  created() {
+    this.updateClaims()
+  },
+  methods: {
+    async updateClaims() {
+      if (this.authState && this.authState.isAuthenticated) {
+        this.claims = await this.$auth.getUser()
+      }
+    },
+    async login() {
+      await this.$auth.signInWithRedirect({originalUri: '/todos'})
+    },
+    async logout() {
+      await this.$auth.signOut()
+    }
+  },
+}
+</script>
+```
+{% endraw %}
+
+This is the top-level component that defines the header bar and includes the router component. The header bar has a login or logout button and will show the authenticated user's email address when logged in.
+
+The app gets the authenticated user's email address from the JWT claims. (A claim is a piece of information asserted about the subject by the authenticating authority.) This happens in the `updateClaims()` method, which is triggered when the component is created, and is also triggered by a watch method so that it is updated as the authenticated state changes.
+
+Create a new file to encapsulate the resource server access logic.
+
+`src/Api.js`
+
+```js
+import axios from 'axios'
+
+const instance = axios.create({
+  baseURL: process.env.VUE_APP_SERVER_URI,
+  timeout: 1000
+});
+
+const createApi = (auth) => {
+
+  instance.interceptors.request.use(async function (config) {
+    let accessToken = await auth.getAccessToken()
+    config.headers = {
+      Authorization: `Bearer ${accessToken}`
+    }
+    return config;
+  }, function (error) {
+    return Promise.reject(error);
+  });
+
+  return {
+
+    // (C)reate
+    createNew(text, completed) {
+      return instance.post('/todos', {title: text, completed: completed})
+    },
+
+    // (R)ead
+    getAll() {
+      return instance.get('/todos', {
+        transformResponse: [function (data) {
+          return data ? JSON.parse(data)._embedded.todos : data;
+        }]
+      })
+    },
+
+    // (U)pdate
+    updateForId(id, text, completed) {
+      return instance.put('todos/' + id, {title: text, completed: completed})
+    },
+
+    // (D)elete
+    removeForId(id) {
+      return instance.delete('todos/' + id)
+    }
+  }
+}
+
+export default createApi
+```
+
+All of the requests to the server go through this module. Notice that the server URI is hard coded at the top as `SERVER_URL`. In production this would be moved to a config file or an environment variable. Also take a look at how the access token is retrieved from the global `auth` object and injected into every request.
```

Comment:
I'm not sure your commend about production makes sense. It's already in a config file and set as an environment variable, isn't it?